crypto: mbedtls_shim: Fix concurrency, leaks, and AEAD output

kaabia opened the pull request:
Fixes <https://github.com/zephyrproject-rtos/zephyr/issues/98956>

This patch addresses critical bugs related to thread safety, resource
management, and correct AEAD behavior in the mbedTLS crypto shim.

1. Concurrency and Deadlock:
   - **Fix session mutex handling:** Ensure the `mtls_sessions_lock` is
     always released in `mtls_get_unused_session_index` on failure
     to prevent deadlocks.
   - **Protect `in_use` flag:** Added mutex protection when setting
     `in_use = false` in free functions.

2. Resource Leaks:
   - **Cleanup on setup failure:** Added calls to `mbedtls_*_free()`
     in `mtls_session_setup()` when key initialization fails.
   - **Free logic fix:** Corrected `if/else` structure in
     `mtls_session_free()` to ensure the correct context is freed.

3. Cipher Integrity:
   - **Fix AEAD output length:** Removed erroneous lines in CCM and GCM
     decryption functions that over-reported the plaintext length.
   - **Logging:** Corrected "encrypt" log messages to "decrypt."

kaabia commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -450,14 +454,17 @@ static int mtls_session_free(const struct device *dev, struct cipher_ctx *ctx)
 
 	if (mtls_session->mode == CRYPTO_CIPHER_MODE_CCM) {
 		mbedtls_ccm_free(&mtls_session->mtls_ccm);
+	}
```
You are absolutely right! Good catch, thank you.

Fixed !

yvesll reviewed (approved):
LGTM, please remeber to update the license header.

ceolin commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -588,7 +594,9 @@ static int mtls_hash_session_free(const struct device *dev, struct hash_ctx *ctx
 	}
+	k_mutex_lock(&mtls_sessions_lock, K_FOREVER);
 	mtls_session->in_use = false;
+	k_mutex_unlock(&mtls_sessions_lock);
```
you are protecting `mtls_session->in_use` here but not doing it in `mtls_session_setup` error handling .. lines 376, 396 ...

kaabia replied:
Good catch! I missed protecting those writes with the mutex in the error paths. I've now applied the lock/unlock around `mtls_sessions[ctx_idx].in_use = false` for consistency and thread safety.

Thank you for the detailed review!

ceolin reviewed (commented):
Changes look good, just check the comment about error handling in `mtls_session_setup`

jhedberg commented:
@ceolin is this something that should be a blocker? I'm not really familiar with this area, however on a general level this looks like it could use splitting into multiple commits.

cfriedt commented:
@ceolin - this will need your approval. It seems like it could be categorized as a release blocker. @valeriosetti - any thoughts?

jhedberg commented:
@kaabia arguably this should be split up into three commits - one for each of your bulletpoints in the description

valeriosetti reviewed (changes requested):
I only find one possible improvement. However being preexisting I can live with it if we say that this PR is mandatory for the 4.3 release.

valeriosetti commented on drivers/crypto/crypto_mtls_shim.c:
```diff
@@ -372,7 +372,10 @@ static int mtls_session_setup(const struct device *dev,
+			k_mutex_lock(&mtls_sessions_lock, K_FOREVER);
 			mtls_sessions[ctx_idx].in_use = false;
+			k_mutex_unlock(&mtls_sessions_lock);
+			mbedtls_aes_free(aes_ctx);
```
Is it really required to have all these `mtls_sessions[ctx_idx].in_use = false;` in this function since it's already done in `mtls_session_free`?
For hashes the session is only freed in `mtls_hash_session_free`

valeriosetti replied:
@ceolin what do you think about this?

kaabia replied:
Hello @valeriosetti,
You are raising an interresting point!
Do you confirm that the mtls_*_session_free is called even when mtls_*_session_setup fails  and not only at the end of a successfull setup !

kaabia committed:
crypto: mbedtls_shim: Fix typos

Corrected "encrypt" log messages to "decrypt."

Signed-off-by: Badr Bacem KAABIA <badrbacemkaabia@gmail.com>

kaabia committed:
crypto: mbedtls_shim: Remove incorrect out_len calculation in AEAD decrypt

Remove erroneous calculation of `apkt->pkt->out_len` in
`mtls_ccm_decrypt_auth` and `mtls_gcm_decrypt_auth`.
The output length should be handled by the caller or the
AEAD decryption process itself.

Signed-off-by: Badr Bacem KAABIA <badrbacemkaabia@gmail.com>

kaabia committed:
crypto: mbedtls_shim: Fix concurrency and deadlock issues

- **Fix session mutex handling:** Ensure the `mtls_sessions_lock` is
    always released in `mtls_get_unused_session_index` on failure
    to prevent deadlocks.
- **Protect `in_use` flag:** Added mutex protection when setting
    `in_use = false` in free functions.
- **Cleanup on setup failure:** Added calls to `mbedtls_*_free()`
    in `mtls_session_setup()` when key initialization fails.
- **Free logic fix:** Corrected `if/else` structure in
    `mtls_session_free()` to ensure the correct context is freed.

Signed-off-by: Badr Bacem KAABIA <badrbacemkaabia@gmail.com>
